### PR TITLE
LOK-2701, LOK-2551 : Store WSMAN Collector Response Metrics in Cortex/Prometheus

### PR DESCRIPTION
**Pull request by MusaidAli**
## Description
Process and Store WSMAN Collector Response Metrics in Cortex/Prometheus.
Initial commit.

## Jira link(s)
- https://opennms.atlassian.net/browse/LOK-2701

## Checklist
* [ ] Follows Lōkahi's [development guidelines.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts.
* [ ] Notify documentation team of any changes to names of screens or features (affects URLs).

### CODE REVIEW

**Comment by MusaidAli** [comment]
Screenshot from 2024-08-14 18-43-54
Metrics are showing from dynamic static json configs.

**Comment by MusaidAli** [comment]
Screenshot from 2024-08-14 18-43-54
Metrics are showing from dynamic static json configs.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/wsmanconfig/Group.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import lombok.Setter;
+import lombok.ToString;
+
+@Getter
+@Setter
+@ToString
+public class Group {
+
+    public ArrayList<Attrib> attrib;
+    public String _name;
```
Comment (naeem-afzal) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/wsmanconfig/Group.java:35]: Group should include:
name, resource-uri, dialect, filter, resource-type

File: metrics-processor/main/src/main/java/org/opennms/horizon/tsdata/collector/TaskSetCollectorWsmanResponseProcessor.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    public TaskSetCollectorWsmanResponseProcessor(CortexTSS cortexTSS, TenantMetricsTracker tenantMetricsTracker) {
+        this.cortexTSS = cortexTSS;
+        this.tenantMetricsTracker = tenantMetricsTracker;
+    }
+
+    public void processWsmanCollectorResponse(String tenantId, String location, TaskResult taskResult)
+            throws IOException {
+        var response = taskResult.getCollectorResponse();
+        Any collectorMetric = response.getResult();
+        Map<String, String> labels = new HashMap<>();
```
Comment (naeem-afzal) [inline metrics-processor/main/src/main/java/org/opennms/horizon/tsdata/collector/TaskSetCollectorWsmanResponseProcessor.java:58]: Make sure all necessary labels are present in response and aligned with monitored entities

File: minion/plugins/shell-commands/src/main/java/org/opennms/horizon/minion/icmp/shell/WSManDetectorCommand.java [hunk 3]
```diff
@@ -102,13 +102,12 @@ public class WSManDetectorCommand implements Action {
 
     @Override
```
Comment (naeem-afzal) [inline minion/plugins/shell-commands/src/main/java/org/opennms/horizon/minion/icmp/shell/WSManDetectorCommand.java:103]: We have dropped WSMAN detector, its no more needed now

**Review by naeem-afzal — COMMENTED** [review]
Also validate collector results accuracy.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java [hunk 4]
```diff
@@ -117,6 +131,9 @@ private TaskDefinition addWsManScannerTask(WsManActiveDiscoveryDTO discovery) {
                             .setStrictSsl(false)
                             .build())
                     .build());
+
+            TaskDefinition collectorTask = addWsManCollectorTask(discovery, host);
```
[the remark this answers is not in the document]
Comment (MusaidAli) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java]: Resolved collector task moved to ScannerResponseService.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/wsmanconfig/Group.java (hunk 1 above)
Comment (MusaidAli) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/wsmanconfig/Group.java:35]: Resolved . added dialect and filter other attributes are there.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java [hunk 5]
```diff
@@ -127,12 +144,77 @@ private TaskDefinition addWsManScannerTask(WsManActiveDiscoveryDTO discovery) {
...
+        String name = String.join("-", "wsman", "collector", "");
+        String id = String.join("-", String.valueOf(discovery.getId()), String.valueOf(discovery.getId()));
+        String taskIdd = identityForWsmanCollectorTask(name, id);
+        return TaskDefinition.newBuilder()
+                .setType(TaskType.COLLECTOR)
+                .setPluginName("WSMANCollector")
+                .setNodeId(discovery.getId())
+                .setId(taskIdd)
+                .setConfiguration(request)
+                .setSchedule(TaskUtils.AZURE_COLLECTOR_SCHEDULE)
```
[the remark this answers is not in the document]
Comment (MusaidAli) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java]: Resolved add ScannerResponseService schedule and set to default schedule.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java [hunk 6]
```diff
@@ -127,12 +144,77 @@ private TaskDefinition addWsManScannerTask(WsManActiveDiscoveryDTO discovery) {
...
+                        .setUsername(discovery.getUserNames())
+                        .setPassword(discovery.getUserPasswords())
+                        .setRetries(TaskUtils.WSMAN_DEFAULT_RETRIES)
+                        .setConnectionTimeout(TaskUtils.WSMAN_DEFAULT_TIMEOUT_MS)
+                        .setMaxElements(5000)
+                        .setReceiveTimeout(TaskUtils.WSMAN_DEFAULT_TIMEOUT_MS)
+                        .setGssAuth(false)
+                        .setStrictSsl(false)
+                        .build())
+                .setSystemDefinition(SystemDefinition.newBuilder()
```
[the remark this answers is not in the document]
Comment (MusaidAli) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java]: For now Added MySQL memory metrics in other static config json file.

[the remark this answers is not in the document]
Comment (MusaidAli) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java]: Now system Definition added dynamic from json file.

File: metrics-processor/main/src/main/java/org/opennms/horizon/tsdata/collector/TaskSetCollectorWsmanResponseProcessor.java (hunk 2 above)
Comment (MusaidAli) [reply metrics-processor/main/src/main/java/org/opennms/horizon/tsdata/collector/TaskSetCollectorWsmanResponseProcessor.java:62]: For now added static will adjust instance and monitored_enity_id dynamic values.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java [hunk 7]
```diff
@@ -71,29 +72,35 @@ public class ScannerTaskSetService {
...
-        taskSetPublisher.publishNewTasks(tenantId, locationId, List.of(task));
+        List<TaskDefinition> tasks = new ArrayList<>();
+        List<TaskDefinition> list = addWsManScannerTask(discovery);
+        taskSetPublisher.publishNewTasks(tenantId, locationId, list);
     }
 
-    private TaskDefinition addWsManScannerTask(WsManActiveDiscoveryDTO discovery) {
+    private List<TaskDefinition> addWsManScannerTask(WsManActiveDiscoveryDTO discovery) {
 
+        String host = "";
```
Comment (smunir-onms) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java:99]: It seems "host" is not being used

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java [hunk 8]
```diff
@@ -195,6 +222,59 @@ private void processWsManScanResponse(String tenantId, Long locationId, WsmanSca
...
     }
 
+    private TaskDefinition addWsManCollectorTask(WsManActiveDiscoveryDTO discovery, String host) {
+
+        URL url = null;
+        WsmanCollectionConfigBean configs = null;
+        try {
+            url = new URL(host);
+        } catch (MalformedURLException e) {
+            System.out.println("The URL is malformed: " + e.getMessage());
```
Comment (smunir-onms) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java:232]: Log this, remove System.out.println

Comment (MusaidAli) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java]: Added log.error

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java [hunk 9]
```diff
@@ -71,29 +72,35 @@ public class ScannerTaskSetService {
...
+        List<TaskDefinition> tasks = new ArrayList<>();
         List<WsmanDetectorRequest> wsmanDetectorRequests = new ArrayList<>();
         for (String ipAddress : discovery.getIpAddressesList()) {
+            host = ipAddress;
```
[the remark this answers is not in the document]
Comment (MusaidAli) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java]: removed,not needed now.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java (hunk 7 above)
Comment (MusaidAli) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java]: removed not being used.

**Comment by MusaidAli** [comment]
Screenshot from 2024-08-19 16-54-49

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java [hunk 10]
```diff
@@ -195,6 +227,65 @@ private void processWsManScanResponse(String tenantId, Long locationId, WsmanSca
...
     }
 
+    private TaskDefinition addWsManCollectorTask(WsManActiveDiscoveryDTO discovery, String host) {
+
+        URL url = null;
+        WsmanCollectionConfigBean configs = null;
+        try {
+            url = new URL(host);
+        } catch (MalformedURLException e) {
+            log.error("The URL is malformed: " + e);
```
Comment (smunir-onms) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java:237]: log.error("The URL is malformed: ", e);

File: shared-lib/protobuf/src/main/proto/wsman.proto [hunk 11]
```diff
@@ -134,5 +144,41 @@ message WsmanConfiguration {
...
+}
+
+message WsmanCollectorResponseMetric {
+  repeated WsmanResultMetric metrics = 1;
+  string group = 2 ;
+}
+
+message WsmanCollectorResponse {
+  bool status = 1;
+  string monitor_type = 2;
```
Comment (cgorantla) [inline shared-lib/protobuf/src/main/proto/wsman.proto:177]: we may not need monitor_type since it is always WSMAN

File: shared-lib/protobuf/src/main/proto/wsman.proto [hunk 12]
```diff
@@ -134,5 +144,41 @@ message WsmanConfiguration {
...
+  uint64 time_stamp = 3;
```
Comment (cgorantla) [inline shared-lib/protobuf/src/main/proto/wsman.proto:178]: This is the only place we need time_stamp, don't need it in WsManResultMetric

File: shared-lib/protobuf/src/main/proto/wsman.proto [hunk 13]
```diff
@@ -134,5 +144,41 @@ message WsmanConfiguration {
...
+  repeated WsmanCollectorResponseMetric group_metrics =4;
+  string instance = 5;
+  string monitored_entity_id = 6;
```
Comment (cgorantla) [inline shared-lib/protobuf/src/main/proto/wsman.proto:181]: don't see `monitored_entity_id` in the request, how do you fill in the response

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java [hunk 14]
```diff
@@ -195,6 +227,65 @@ private void processWsManScanResponse(String tenantId, Long locationId, WsmanSca
...
+        }
+        try {
+            URL resourceURL = this.getClass().getResource("/microsoft-windows.json");
+            configs = objectMapper.readValue(resourceURL, WsmanCollectionConfigBean.class);
+        } catch (IOException ex) {
+            throw new RuntimeException(ex);
+        }
+
+        assert url != null;
+        assert configs != null;
```
Comment (cgorantla) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java:247]: why do you need to use assert here ?

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java [hunk 15]
```diff
@@ -195,6 +227,65 @@ private void processWsManScanResponse(String tenantId, Long locationId, WsmanSca
...
+        WsmanCollectionConfigBean configs = null;
+        try {
+            url = new URL(host);
+        } catch (MalformedURLException e) {
+            log.error("The URL is malformed: ", e);
+        }
+        try {
+            URL resourceURL = this.getClass().getResource("/microsoft-windows.json");
+            configs = objectMapper.readValue(resourceURL, WsmanCollectionConfigBean.class);
+        } catch (IOException ex) {
```
Comment (cgorantla) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java:242]: add error log here.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java [hunk 16]
```diff
@@ -195,6 +227,65 @@ private void processWsManScanResponse(String tenantId, Long locationId, WsmanSca
...
+            throw new RuntimeException(ex);
```
Comment (cgorantla) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java:243]: probably need to convert this to a checked exception and handle gracefully. Unlikely that this would ever occur since it is in the source.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java [hunk 17]
```diff
@@ -195,6 +227,65 @@ private void processWsManScanResponse(String tenantId, Long locationId, WsmanSca
...
+
+        URL url = null;
+        WsmanCollectionConfigBean configs = null;
+        try {
+            url = new URL(host);
+        } catch (MalformedURLException e) {
+            log.error("The URL is malformed: ", e);
+        }
+        try {
+            URL resourceURL = this.getClass().getResource("/microsoft-windows.json");
```
Comment (cgorantla) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java:240]: Add a comment above, as of now we assume it to be default windows config.

File: shared-lib/protobuf/src/main/proto/wsman.proto (hunk 11 above)
Comment (MusaidAli) [reply shared-lib/protobuf/src/main/proto/wsman.proto]: Yes,it is removed.

File: shared-lib/protobuf/src/main/proto/wsman.proto (hunk 13 above)
Comment (MusaidAli) [reply shared-lib/protobuf/src/main/proto/wsman.proto]: we have just added static values ,just to trigger request to minion and naeem is working on monitored_entities pr will update according.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java (hunk 14 above)
Comment (MusaidAli) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java]: Removed not needed.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java (hunk 15 above)
Comment (MusaidAli) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java]: added log and checked exception

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java (hunk 16 above)
Comment (MusaidAli) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java]: Resolved added checked exception.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java (hunk 17 above)
Comment (MusaidAli) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java]: added comment 
  // default windows host config

File: shared-lib/protobuf/src/main/proto/wsman.proto (hunk 12 above)
Comment (MusaidAli) [reply shared-lib/protobuf/src/main/proto/wsman.proto]: Removed from other places.

**Comment by MusaidAli** [comment]
These metric labels are tested and works fine.

File: minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanCollectorManager.java [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+                                          *
+                                          *      https://www.gnu.org/licenses/agpl-3.0.txt
+                                          *
+                                          * Unless required by applicable law or agreed to in writing,
+                                          * software distributed under the License is distributed on an
+                                          * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+                                          * either express or implied.  See the License for the specific
+                                          * language governing permissions and limitations under the
+                                          * License.
+                                          */
```
Comment (cgorantla) [inline minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanCollectorManager.java:42]: remove

File: rest-server/src/main/java/org/opennms/horizon/server/service/grpc/InventoryClient.java [hunk 19]
```diff
@@ -27,9 +27,6 @@
 import io.grpc.ManagedChannel;
```
Comment (cgorantla) [inline rest-server/src/main/java/org/opennms/horizon/server/service/grpc/InventoryClient.java:27]: unnecessary change

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java [hunk 20]
```diff
@@ -198,6 +227,70 @@ private void processWsManScanResponse(String tenantId, Long locationId, WsmanSca
...
+        for (org.opennms.horizon.inventory.discovery.BasicAuth ba :
+                discovery.getCredentials().getBasicAuthList()) {
+            crBuilder.addBasicAuth(BasicAuth.newBuilder()
+                    .setUsername(ba.getUsername())
+                    .setPassword(ba.getPassword())
+                    .build());
+        }
+        Map<String, String> labels = new HashMap<>();
+        labels.put("node_id", Long.toString(node.getId()));
+        labels.put("ip_address", host);
```
Comment (cgorantla) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java:255]: just use `hostname` here.

File: minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanCollector.java [hunk 21]
```diff
@@ -0,0 +1,292 @@
...
+        @Override
+        public Thread newThread(Runnable r) {
+            return new Thread(r, "Wsman-Proxy-RPC-Session-Reaper");
+        }
+    });
+
+    @Override
+    public CompletableFuture<CollectionSet> collect(CollectionRequest request, Any config) {
+
+        LOG.debug("WSMAN Collector invoker ..... ");
```
Comment (cgorantla) [inline minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanCollector.java:79]: unnecessary debug statement

File: minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanCollector.java [hunk 22]
```diff
@@ -0,0 +1,292 @@
...
+        try {
+            if (!config.is(WsmanCollectorRequest.class)) {
+                throw new IllegalArgumentException(
+                        "config must be an WsmanCollectorRequest; type-url=" + config.getTypeUrl());
+            }
+
+            wsmanCollectorRequest = config.unpack(WsmanCollectorRequest.class);
+            WsmanCollectorRequest finalWsmanCollectorRequest = wsmanCollectorRequest;
+
+            return CompletableFuture.supplyAsync(
```
Comment (cgorantla) [inline minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanCollector.java:90]: this part of code still doesn't look readable or easy to review. If you can improve this without breaking functionality that would be great.

File: shared-lib/protobuf/src/main/proto/wsman.proto [hunk 23]
```diff
@@ -134,5 +144,41 @@ message WsmanConfiguration {
...
+  string group = 2 ;
+}
+
+message WsmanCollectorResponse {
+  bool status = 1;
+  uint64 time_stamp = 2;
+  repeated WsmanCollectorResponseMetric group_metrics =3;
+  string instance = 4;
+  uint64 node_id = 5;
+  map<string, string> metric_labels = 6;
```
Comment (cgorantla) [inline shared-lib/protobuf/src/main/proto/wsman.proto:181]: I think this is unused, not being filled anywhere.

Comment (cgorantla) [inline shared-lib/protobuf/src/main/proto/wsman.proto:181]: you have to add this to CollectorResponse message in proto and fill it [here](https://github.com/OpenNMS-Cloud/lokahi/blob/1e50062b6d8ccf058ab0ccca476d912574d7a68a/minion/minion-taskset/taskset-worker/src/main/java/org/opennms/horizon/minion/taskset/worker/impl/TaskExecutionResultProcessorImpl.java#L176)

File: minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanCollectorManager.java (hunk 18 above)
Comment (MusaidAli) [reply minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanCollectorManager.java:42]: Removed ,was merging develop issue.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java (hunk 20 above)
Comment (MusaidAli) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/response/ScannerResponseService.java:255]: Added hostname.

File: minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanCollector.java (hunk 21 above)
Comment (MusaidAli) [reply minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanCollector.java:79]: Removed.

File: shared-lib/protobuf/src/main/proto/wsman.proto (hunk 23 above)
Comment (MusaidAli) [reply shared-lib/protobuf/src/main/proto/wsman.proto]: I think it is not being used and removed ,metric labels are already passing will go with this flow ,It will be then complicated.Thanks

Comment (MusaidAli) [reply shared-lib/protobuf/src/main/proto/wsman.proto]: Yes,not being used and removed.

**Comment by MusaidAli** [comment]
Screenshot from 2024-08-29 14-38-44
hostname label added.

File: minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanCollector.java (hunk 22 above)
Comment (MusaidAli) [reply minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanCollector.java:89]: using variables in lambda expression should be final ,it is being created by Intellij.

File: rest-server/src/main/java/org/opennms/horizon/server/service/grpc/InventoryClient.java (hunk 19 above)
Comment (MusaidAli) [reply rest-server/src/main/java/org/opennms/horizon/server/service/grpc/InventoryClient.java:27]: was due to develop merge .

**Comment by MusaidAli** [comment]
Added label hostname .

Screenshot from 2024-08-29 17-16-06

**Review by cgorantla — APPROVED** [review]
lgtm!
